### ISSUE TITLE
Support Wicket dynamic Ajax endpoints

### ISSUE THREAD

**Issue by mdindoffer** [elastic/apm-agent-java#348]
We use the Wicket framework for an AJAX-heavy web application.
To utilize the AJAX requests in Wicket, custom Ajax handlers are dynamically generated and attached to components, such as:
```java
childComponent.add( new OnChangeAjaxBehavior() {
    @Override
    protected void onUpdate( AjaxRequestTarget target ) {
        //Handle AJAX call
    }
}
```
Wicket then generates an AJAX URL similar to:
`/page?4-1.IBehaviorListener.0-parent~component-child~component`

As a result, the APM agent merges all AJAX requests on all pages into one generic POST request type:
![screenshot_2018-11-26 transactions](https://user-images.githubusercontent.com/12681399/49009873-ee839780-f171-11e8-98c7-26072d821946.png)

It would be great if the APM agent knew how to deal with this and categorize the requests properly, but I'm not sure if it is even possible.
Maybe there's a workaround?

**Comment by felixbarny** [comment]
One workaround would be to use the public API and add `ElasticApm.currentTransaction().setName("update cart quantity")`, or similar to your ajax handlers. But I see that this would be quite a lot of effort to name everything by hand. How would you like the transaction name to be ideally?

**Comment by eyalkoren** [comment]
@mdindoffer can [this](https://www.elastic.co/guide/en/apm/agent/java/current/config-http.html#config-use-path-as-transaction-name) together with [this](https://www.elastic.co/guide/en/apm/agent/java/current/config-http.html#config-url-groups) work?

**Comment by felixbarny** [comment]
I don't think so because the important part URL `/page?4-1.IBehaviorListener.0-parent~component-child~component` is not in the path but in the query parameters.

**Comment by mdindoffer** [comment]
Well, if the entire URL was captured in the path (haven't tried yet), it might be a solution.

The ideal unique identifier / transaction name would capture the URL of the page plus the component path (the URL query) without the versioning numbers. For example this URL:
`http://host:8080/example/resource/edit/123?6-1.IBehaviorListener.0-resource~form-detail~panel-somePropertyList`
would be converted to a transaction:
`resource/edit/IBehaviorListener.0-resource~form-detail~panel-somePropertyList`
i.e. strip the `123` resource id and `6-1` Wicket versioning scheme at the beginning.

If I understand correctly, the `url_groups` property does not parse regex, only allows for simple * wildcard, so we cannot filter by number scheme. Anyway, I *think* two url group patterns could solve 80% of our use cases:
1. `?*-*.I`
2. */edit/*

However, the pattern nr.2 would have to be somehow applied **only** to the URL part, not the query params, since it would swallow them.

Is my reasoning OK ?

**Comment by felixbarny** [comment]
I think the best and easiest way would be for you to create a `javax.servlet.Filter`, get the parameters of the request and see if there is one which matches `.*I.*Listener.*`. If so, construct an appropriate name based on that and use the public API to set `ElasticApm.currentTransaction().setName(wicketAjaxTransactionName)`.

**Comment by felixbarny** [comment]
If you have something which works nicely I think it would be great if you could share that with the community :)

Maybe as a [gist](http://gist.github.com).

**Comment by mdindoffer** [comment]
Hi Felix. The servlet filter workaround works nicely :)
Here's my solution:
https://gist.github.com/mdindoffer/21d61433e1cd2b3c23c5880c76a11713

A couple of important points I found out while implementing the filter:
1. Full wicket component paths may easily surpass 100 or so characters making them unusable as the transaction names, so I've reduced the paths to only the listener and the target component's identifier.
2. **Warning**: Kibana has some weird stuff going on! The APM UI uses an in-house URI conversion process, where the `'%'` characters are encoded as `'~'` on a webpage. Clicking a link in the Transaction UI then calls a simple `.replace(/~/g, "%")` JS function before parsing the URL. Since Wicket paths already do contain tilde characters, one cannot use them in raw form as the transaction link will result in a malformed URI. As a workaround I've replaced the Wicket tildes with an underscore.

I feel the point nr.2 should be mentioned in the APM docs so people won't accidentally name their transactions with ~ in them.

**Comment by sqren** [comment]
> TLDR; having a `~` in a transaction name causes malformed URLs

Yes :(

> **Warning**: Kibana has some weird stuff going on! The APM UI uses an in-house URI conversion process, where the `'%'` characters are encoded as `'~'` on a webpage. Clicking a link in the Transaction UI then calls a simple `.replace(/~/g, "%")` JS function before parsing the URL. Since Wicket paths already do contain tilde characters, one cannot use them in raw form as the transaction link will result in a malformed URI. As a workaround I've replaced the Wicket tildes with an underscore.

Well spotted, and glad you were able to add a work-around. This is definitely not optimal, and we already have [an issue](https://github.com/elastic/kibana/issues/24892) open about this that [explains why it is so](https://github.com/elastic/kibana/issues/24892#issuecomment-434876263). I'll try to bump the priority so we can get this fixed.

**Comment by sqren** [comment]
fyi @felixbarny @mdindoffer The issue related to weird encoding in APM UI (Kibana) has finally been fixed :)
https://github.com/elastic/kibana/pull/33309

**Comment by SylvainJuge** [comment]
Closing this issue as a work-around is available in [apm-contrib](https://github.com/elastic/apm-contrib/tree/master/apm-agent-java/frameworks/wicket) repository.
We will re-evaluate adding native support for this in the agent if there are more requests for it.
